File: go/fast_shift_test.go
package pragmastat

import (
	"math"
	"math/rand"
	"sort"
	"testing"
)

func naiveShift[T Number](x, y []T) float64 {
	var diffs []float64
	for _, xi := range x {
		for _, yj := range y {
			diffs = append(diffs, float64(xi-yj))
		}
	}
	sort.Float64s(diffs)

	n := len(diffs)
	if n%2 == 0 {
		return (diffs[n/2-1] + diffs[n/2]) / 2.0
	}
	return diffs[n/2]
}

func TestFastShiftSmallArrays(t *testing.T) {
	const tolerance = 1e-9
	rng := rand.New(rand.NewSource(1729))

	for m := 1; m <= 20; m++ {
		for n := 1; n <= 20; n++ {
			for iteration := 0; iteration < 3; iteration++ {
				x := make([]float64, m)
				y := make([]float64, n)
				for i := range x {
					x[i] = rng.NormFloat64()
				}
				for i := range y {
					y[i] = rng.NormFloat64()
				}

				actual, err := fastShift(x, y)
				if err != nil {
					t.Errorf("fastShift(%d, %d) error: %v", m, n, err)
					continue
				}

				expected := naiveShift(x, y)
				if math.Abs(actual-expected) > tolerance {
					t.Errorf("fastShift(%d, %d) = %v, want %v", m, n, actual, expected)
				}
			}
		}
	}
}

func TestFastShiftMediumArrays(t *testing.T) {
	const tolerance = 1e-9
	rng := rand.New(rand.NewSource(42))

	for size := 20; size <= 100; size += 10 {
		for iteration := 0; iteration < 3; iteration++ {
			x := make([]float64, size)
			y := make([]float64, size/2)
			for i := range x {
				x[i] = rng.NormFloat64()
			}
			for i := range y {
				y[i] = rng.NormFloat64()
			}

			actual, err := fastShift(x, y)
			if err != nil {
				t.Errorf("fastShift(%d, %d) error: %v", size, size/2, err)
				continue
			}

			expected := naiveShift(x, y)
			if math.Abs(actual-expected) > tolerance {
				t.Errorf("fastShift(%d, %d) = %v, want %v", size, size/2, actual, expected)
			}
		}
	}
}

func TestFastShiftDifferentDistributions(t *testing.T) {
	const tolerance = 1e-9
	seed := int64(2024)

	type distribution struct {
		name  string
		mean  float64
		scale float64
	}

	distributions := []distribution{
		{"standard", 0.0, 1.0},
		{"shifted", 5.0, 2.0},
		{"negative", -10.0, 1.0},
		{"small", 0.0, 0.1},
	}

	for _, dist := range distributions {
		t.Run(dist.name, func(t *testing.T) {
			rng := rand.New(rand.NewSource(seed))
			seed++

			for trial := 0; trial < 10; trial++ {
				x := make([]float64, 15)
				y := make([]float64, 10)
				for i := range x {
					x[i] = dist.mean + dist.scale*rng.NormFloat64()
				}
				for i := range y {
					y[i] = dist.mean + dist.scale*rng.NormFloat64()
				}

				actual, err := fastShift(x, y)
				if err != nil {
					t.Errorf("fastShift error: %v", err)
					continue
				}

				expected := naiveShift(x, y)
				if math.Abs(actual-expected) > tolerance {
					t.Errorf("fastShift = %v, want %v", actual, expected)
				}
			}
		})
	}
}

func TestFastShiftSingleElement(t *testing.T) {
	const tolerance = 1e-9
	rng := rand.New(rand.NewSource(123))

	for trial := 0; trial < 20; trial++ {
		x := []float64{rng.NormFloat64()}
		y := []float64{rng.NormFloat64()}

		result, err := fastShift(x, y)
		if err != nil {
			t.Errorf("fastShift single element error: %v", err)
			continue
		}

		expected := x[0] - y[0]
		if math.Abs(result-expected) > tolerance {
			t.Errorf("fastShift([%v], [%v]) = %v, want %v", x[0], y[0], result, expected)
		}
	}
}

func TestFastShiftIdenticalArrays(t *testing.T) {
	const tolerance = 1e-9
	rng := rand.New(rand.NewSource(456))

	for size := 1; size <= 30; size++ {
		for trial := 0; trial < 3; trial++ {
			x := make([]float64, size)
			for i := range x {
				x[i] = rng.NormFloat64()
			}

			result, err := fastShift(x, x)
			if err != nil {
				t.Errorf("fastShift identical arrays error: %v", err)
				continue
			}

			if math.Abs(result) > tolerance {
				t.Errorf("fastShift(x, x) = %v, want 0.0", result)
			}
		}
	}
}

func TestFastShiftAsymmetricSizes(t *testing.T) {
	const tolerance = 1e-9
	rng := rand.New(rand.NewSource(789))

	configs := []struct{ m, n int }{
		{1, 100},
		{100, 1},
		{10, 50},
		{50, 10},
		{5, 200},
	}

	for _, cfg := range configs {
		t.Run("", func(t *testing.T) {
			x := make([]float64, cfg.m)
			y := make([]float64, cfg.n)
			for i := range x {
				x[i] = rng.NormFloat64()
			}
			for i := range y {
				y[i] = rng.NormFloat64()
			}

			actual, err := fastShift(x, y)
			if err != nil {
				t.Errorf("fastShift(%d, %d) error: %v", cfg.m, cfg.n, err)
				return
			}

			expected := naiveShift(x, y)
			if math.Abs(actual-expected) > tolerance {
				t.Errorf("fastShift(%d, %d) = %v, want %v", cfg.m, cfg.n, actual, expected)
			}
		})
	}
}

func TestFastShiftExtremeValues(t *testing.T) {
	const tolerance = 1e-9

	tests := []struct {
		name string
		x    []float64
		y    []float64
	}{
		{"large values", []float64{1e6, 2e6, 3e6}, []float64{1e5, 2e5}},
		{"small values", []float64{1e-8, 2e-8, 3e-8}, []float64{1e-9, 2e-9}},
		{"negative values", []float64{-100, -50, -10}, []float64{-80, -40}},
		{"mixed values", []float64{-100, 0, 100}, []float64{-50, 50}},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			actual, err := fastShift(tt.x, tt.y)
			if err != nil {
				t.Errorf("fastShift error: %v", err)
				return
			}

			expected := naiveShift(tt.x, tt.y)
			if math.Abs(actual-expected) > tolerance {
				t.Errorf("fastShift = %v, want %v", actual, expected)
			}

			if math.IsNaN(actual) {
				t.Errorf("fastShift returned NaN")
			}
			if math.IsInf(actual, 0) {
				t.Errorf("fastShift returned Inf")
			}
		})
	}
}

func TestFastShiftDuplicateValues(t *testing.T) {
	const tolerance = 1e-9
	rng := rand.New(rand.NewSource(222))

	for trial := 0; trial < 10; trial++ {
		x := make([]float64, 12)
		y := make([]float64, 10)
		for i := range x {
			x[i] = math.Round(rng.NormFloat64()*5) / 5.0
		}
		for i := range y {
			y[i] = math.Round(rng.NormFloat64()*5) / 5.0
		}

		actual, err := fastShift(x, y)
		if err != nil {
			t.Errorf("fastShift with duplicates error: %v", err)
			continue
		}

		expected := naiveShift(x, y)
		if math.Abs(actual-expected) > tolerance {
			t.Errorf("fastShift = %v, want %v", actual, expected)
		}
	}
}

func TestFastShiftZeroSpread(t *testing.T) {
	const tolerance = 1e-9

	x := []float64{5.0, 5.0, 5.0, 5.0, 5.0}
	y := []float64{2.0, 2.0, 2.0, 2.0}

	result, err := fastShift(x, y)
	if err != nil {
		t.Errorf("fastShift zero spread error: %v", err)
		return
	}

	expected := 3.0
	if math.Abs(result-expected) > tolerance {
		t.Errorf("fastShift = %v, want %v", result, expected)
	}
}

func TestFastShiftShiftInvariance(t *testing.T) {
	const tolerance = 1e-6
	rng := rand.New(rand.NewSource(555))

	for trial := 0; trial < 10; trial++ {
		x := make([]float64, 15)
		y := make([]float64, 12)
		for i := range x {
			x[i] = rng.NormFloat64()
		}
		for i := range y {
			y[i] = rng.NormFloat64()
		}
		shift := rng.NormFloat64() * 10

		result1, _ := fastShift(x, y)

		xShifted := make([]float64, len(x))
		for i := range x {
			xShifted[i] = x[i] + shift
		}
		result2, _ := fastShift(xShifted, y)

		if math.Abs(result2-(result1+shift)) > tolerance {
			t.Errorf("X shift invariance failed: got %v, want %v", result2, result1+shift)
		}

		yShifted := make([]float64, len(y))
		for i := range y {
			yShifted[i] = y[i] + shift
		}
		result3, _ := fastShift(x, yShifted)

		if math.Abs(result3-(result1-shift)) > tolerance {
			t.Errorf("Y shift invariance failed: got %v, want %v", result3, result1-shift)
		}
	}
}

func TestFastShiftScaleInvariance(t *testing.T) {
	const tolerance = 1e-6
	rng := rand.New(rand.NewSource(777))

	for trial := 0; trial < 10; trial++ {
		x := make([]float64, 15)
		y := make([]float64, 12)
		for i := range x {
			x[i] = rng.NormFloat64()
		}
		for i := range y {
			y[i] = rng.NormFloat64()
		}

		result1, _ := fastShift(x, y)

		scale := 2.0
		xScaled := make([]float64, len(x))
		yScaled := make([]float64, len(y))
		for i := range x {
			xScaled[i] = x[i] * scale
		}
		for i := range y {
			yScaled[i] = y[i] * scale
		}

		result2, _ := fastShift(xScaled, yScaled)

		if math.Abs(result2-result1*scale) > tolerance {
			t.Errorf("Scale invariance failed: got %v, want %v", result2, result1*scale)
		}
	}
}

func TestFastShiftErrorHandling(t *testing.T) {
	tests := []struct {
		name string
		x    []float64
		y    []float64
	}{
		{"empty x", []float64{}, []float64{1.0}},
		{"empty y", []float64{1.0}, []float64{}},
		{"both empty", []float64{}, []float64{}},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			_, err := fastShift(tt.x, tt.y)
			if err == nil {
				t.Errorf("fastShift(%v, %v) expected error, got nil", tt.x, tt.y)
			}
		})
	}
}

func TestFastShiftIntegerInputs(t *testing.T) {
	const tolerance = 1e-9

	tests := []struct {
		name string
		x    []int
		y    []int
	}{
		{"simple", []int{3, 5, 7}, []int{1, 2}},
		{"negative", []int{-5, -3, -1}, []int{1, 2}},
		{"single", []int{10}, []int{3}},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result, err := fastShift(tt.x, tt.y)
			if err != nil {
				t.Errorf("fastShift error: %v", err)
				return
			}

			expected := naiveShift(tt.x, tt.y)
			if math.Abs(result-expected) > tolerance {
				t.Errorf("fastShift(%v, %v) = %v, want %v", tt.x, tt.y, result, expected)
			}
		})
	}
}

func BenchmarkFastShift(b *testing.B) {
	rng := rand.New(rand.NewSource(1729))

	sizes := []struct{ m, n int }{
		{100, 100},
		{200, 200},
		{500, 500},
		{1000, 1000},
	}

	for _, size := range sizes {
		x := make([]float64, size.m)
		y := make([]float64, size.n)
		for i := range x {
			x[i] = rng.NormFloat64()
		}
		for i := range y {
			y[i] = rng.NormFloat64()
		}

		b.Run("", func(b *testing.B) {
			for i := 0; i < b.N; i++ {
				_, _ = fastShift(x, y)
			}
		})
	}
}

func BenchmarkNaiveShift(b *testing.B) {
	rng := rand.New(rand.NewSource(1729))

	sizes := []struct{ m, n int }{
		{100, 100},
		{200, 200},
	}

	for _, size := range sizes {
		x := make([]float64, size.m)
		y := make([]float64, size.n)
		for i := range x {
			x[i] = rng.NormFloat64()
		}
		for i := range y {
			y[i] = rng.NormFloat64()
		}

		b.Run("", func(b *testing.B) {
			for i := 0; i < b.N; i++ {
				_ = naiveShift(x, y)
			}
		})
	}
}

File: ts/tests/fastShift.test.ts
import { fastShift } from '../src/fastShift';

const TOLERANCE = 1e-9;

function generateRandomValues(n: number, seed: number): number[] {
  const values: number[] = [];
  let state = seed;
  for (let i = 0; i < n; i++) {
    state = (state * 1664525 + 1013904223) % 4294967296;
    values.push((state / 4294967296 - 0.5) * 10);
  }
  return values;
}

function naiveQuantiles(x: number[], y: number[], p: number[]): number[] {
  const diffs: number[] = [];
  for (const xi of x) {
    for (const yj of y) {
      diffs.push(xi - yj);
    }
  }
  diffs.sort((a, b) => a - b);

  const result: number[] = [];
  for (const pk of p) {
    const n = diffs.length;
    const h = 1.0 + (n - 1) * pk;
    const lo = Math.max(1, Math.min(n, Math.floor(h)));
    const hi = Math.max(1, Math.min(n, Math.ceil(h)));
    const gamma = h - lo;

    const a = diffs[lo - 1];
    const b = diffs[hi - 1];

    result.push(gamma === 0.0 ? a : (1.0 - gamma) * a + gamma * b);
  }
  return result;
}

describe('fastShift', () => {
  describe('Small arrays - matches naive', () => {
    it('should match naive for various small array sizes', () => {
      const seed = 1729;
      let currentSeed = seed;

      for (let m = 1; m <= 20; m++) {
        for (let n = 1; n <= 20; n++) {
          for (let iteration = 0; iteration < 2; iteration++) {
            const x = generateRandomValues(m, currentSeed++);
            const y = generateRandomValues(n, currentSeed++);
            const p = [0.0, 0.25, 0.5, 0.75, 1.0];

            const actual = fastShift(x, y, p);
            const expected = naiveQuantiles(x, y, p);

            expect(actual.length).toBe(expected.length);
            for (let i = 0; i < expected.length; i++) {
              expect(actual[i]).toBeCloseTo(expected[i], 9);
            }
          }
        }
      }
    });
  });

  describe('Medium arrays - matches naive', () => {
    it('should match naive for medium-sized arrays', () => {
      const seed = 42;
      let currentSeed = seed;

      for (let size = 20; size <= 100; size += 20) {
        for (let iteration = 0; iteration < 2; iteration++) {
          const x = generateRandomValues(size, currentSeed++);
          const y = generateRandomValues(Math.floor(size / 2), currentSeed++);
          const p = [0.1, 0.5, 0.9];

          const actual = fastShift(x, y, p);
          const expected = naiveQuantiles(x, y, p);

          expect(actual.length).toBe(expected.length);
          for (let i = 0; i < expected.length; i++) {
            expect(actual[i]).toBeCloseTo(expected[i], 9);
          }
        }
      }
    });
  });

  describe('Different distributions - all quantiles', () => {
    it('should work correctly with different distributions', () => {
      const distributions = [
        { mean: 0, scale: 1 },
        { mean: 5, scale: 2 },
        { mean: -10, scale: 1 },
        { mean: 0, scale: 0.1 },
      ];

      const probabilities = [0.0, 0.05, 0.1, 0.25, 0.5, 0.75, 0.9, 0.95, 1.0];

      let seed = 2024;
      for (const dist of distributions) {
        const x = generateRandomValues(15, seed++).map((v) => v * dist.scale + dist.mean);
        const y = generateRandomValues(10, seed++).map((v) => v * dist.scale + dist.mean);

        const actual = fastShift(x, y, probabilities);
        const expected = naiveQuantiles(x, y, probabilities);

        expect(actual.length).toBe(expected.length);
        for (let i = 0; i < expected.length; i++) {
          expect(actual[i]).toBeCloseTo(expected[i], 9);
        }
      }
    });
  });

  describe('Unsorted input - matches sorted', () => {
    it('should handle unsorted input correctly', () => {
      const seed = 999;
      let currentSeed = seed;

      for (let trial = 0; trial < 20; trial++) {
        const xRaw = generateRandomValues(20, currentSeed++);
        const yRaw = generateRandomValues(15, currentSeed++);
        const p = [0.25, 0.5, 0.75];

        const xSorted = [...xRaw].sort((a, b) => a - b);
        const ySorted = [...yRaw].sort((a, b) => a - b);

        const xShuffled = [...xRaw].sort(() => Math.random() - 0.5);
        const yShuffled = [...yRaw].sort(() => Math.random() - 0.5);

        const resultUnsorted = fastShift(xShuffled, yShuffled, p, false);
        const resultSorted = fastShift(xSorted, ySorted, p, true);

        expect(resultUnsorted.length).toBe(resultSorted.length);
        for (let i = 0; i < resultSorted.length; i++) {
          expect(resultUnsorted[i]).toBeCloseTo(resultSorted[i], 9);
        }
      }
    });
  });

  describe('Single element - returns constant', () => {
    it('should return constant difference for single elements', () => {
      const seed = 123;
      let currentSeed = seed;

      for (let trial = 0; trial < 10; trial++) {
        const x = generateRandomValues(1, currentSeed++);
        const y = generateRandomValues(1, currentSeed++);
        const p = [0.0, 0.25, 0.5, 0.75, 1.0];

        const result = fastShift(x, y, p);
        const expected = x[0] - y[0];

        for (const q of result) {
          expect(q).toBeCloseTo(expected, 9);
        }
      }
    });
  });

  describe('Identical arrays - median is zero', () => {
    it('should return zero median for identical arrays', () => {
      const seed = 456;

      for (let size = 1; size <= 30; size += 5) {
        const x = generateRandomValues(size, seed + size);
        const p = [0.5];

        const result = fastShift(x, x, p);

        expect(result[0]).toBeCloseTo(0.0, 9);
      }
    });
  });

  describe('Asymmetric sizes - correct results', () => {
    it('should handle asymmetric array sizes correctly', () => {
      const configs = [
        { m: 1, n: 100 },
        { m: 100, n: 1 },
        { m: 10, n: 50 },
        { m: 50, n: 10 },
      ];

      const seed = 789;
      let currentSeed = seed;

      for (const { m, n } of configs) {
        const x = generateRandomValues(m, currentSeed++);
        const y = generateRandomValues(n, currentSeed++);
        const p = [0.0, 0.5, 1.0];

        const actual = fastShift(x, y, p);
        const expected = naiveQuantiles(x, y, p);

        expect(actual.length).toBe(expected.length);
        for (let i = 0; i < expected.length; i++) {
          expect(actual[i]).toBeCloseTo(expected[i], 9);
        }
      }
    });
  });

  describe('Extreme quantiles - match min/max', () => {
    it('should match actual min and max for extreme quantiles', () => {
      const seed = 321;
      let currentSeed = seed;

      for (let trial = 0; trial < 10; trial++) {
        const x = generateRandomValues(10 + trial, currentSeed++);
        const y = generateRandomValues(8 + Math.floor(trial / 2), currentSeed++);
        const p = [0.0, 1.0];

        const result = fastShift(x, y, p);

        let min = Infinity;
        let max = -Infinity;
        for (const xi of x) {
          for (const yj of y) {
            const diff = xi - yj;
            if (diff < min) min = diff;
            if (diff > max) max = diff;
          }
        }

        expect(result[0]).toBeCloseTo(min, 9);
        expect(result[1]).toBeCloseTo(max, 9);
      }
    });
  });

  describe('Many probabilities - monotonic increasing', () => {
    it('should produce monotonically increasing quantiles', () => {
      const seed = 654;
      let currentSeed = seed;

      for (let trial = 0; trial < 10; trial++) {
        const x = generateRandomValues(25, currentSeed++);
        const y = generateRandomValues(20, currentSeed++);

        const p = Array.from({ length: 21 }, (_, i) => i / 20.0);

        const result = fastShift(x, y, p);

        for (let i = 1; i < result.length; i++) {
          expect(result[i]).toBeGreaterThanOrEqual(result[i - 1] - TOLERANCE);
        }
      }
    });
  });

  describe('Negative values - handled correctly', () => {
    it('should handle negative values correctly', () => {
      const seed = 111;

      for (let trial = 0; trial < 10; trial++) {
        const x = generateRandomValues(15, seed + trial).map((v) => v - 50);
        const y = generateRandomValues(12, seed + trial + 1000).map((v) => v - 50);
        const p = [0.25, 0.5, 0.75];

        const actual = fastShift(x, y, p);
        const expected = naiveQuantiles(x, y, p);

        for (let i = 0; i < expected.length; i++) {
          expect(actual[i]).toBeCloseTo(expected[i], 9);
        }
      }
    });
  });

  describe('Duplicate values - handled correctly', () => {
    it('should handle arrays with duplicate values', () => {
      const seed = 222;

      for (let trial = 0; trial < 5; trial++) {
        const x = generateRandomValues(12, seed + trial).map((v) => Math.round(v * 5) / 5.0);
        const y = generateRandomValues(10, seed + trial + 500).map((v) => Math.round(v * 5) / 5.0);
        const p = [0.0, 0.5, 1.0];

        const actual = fastShift(x, y, p);
        const expected = naiveQuantiles(x, y, p);

        for (let i = 0; i < expected.length; i++) {
          expect(actual[i]).toBeCloseTo(expected[i], 9);
        }
      }
    });
  });

  describe('Very small values - numerical stability', () => {
    it('should maintain numerical stability with very small values', () => {
      const seed = 333;

      for (let trial = 0; trial < 5; trial++) {
        const x = generateRandomValues(10, seed + trial).map((v) => v * 1e-8);
        const y = generateRandomValues(10, seed + trial + 200).map((v) => v * 1e-8);
        const p = [0.5];

        const result = fastShift(x, y, p);

        expect(result[0]).not.toBeNaN();
        expect(isFinite(result[0])).toBe(true);
      }
    });
  });

  describe('Large values - numerical stability', () => {
    it('should maintain numerical stability with large values', () => {
      const seed = 444;

      for (let trial = 0; trial < 5; trial++) {
        const x = generateRandomValues(10, seed + trial).map((v) => v * 1e6 + 1e6);
        const y = generateRandomValues(10, seed + trial + 300).map((v) => v * 1e5 + 1e6);
        const p = [0.5];

        const result = fastShift(x, y, p);

        expect(result[0]).not.toBeNaN();
        expect(isFinite(result[0])).toBe(true);
      }
    });
  });

  describe('Zero spread - all same', () => {
    it('should return constant when all elements are same', () => {
      const x = Array(10).fill(5.0);
      const y = Array(8).fill(2.0);
      const p = [0.0, 0.25, 0.5, 0.75, 1.0];

      const result = fastShift(x, y, p);

      for (const q of result) {
        expect(q).toBeCloseTo(3.0, 9);
      }
    });
  });

  describe('Large arrays - performance test', () => {
    it('should complete in reasonable time for 500x500 arrays', () => {
      const x = generateRandomValues(500, 1729);
      const y = generateRandomValues(500, 1730);
      const p = [0.5];

      const start = Date.now();
      const result = fastShift(x, y, p);
      const elapsed = Date.now() - start;

      expect(result.length).toBe(1);
      expect(elapsed).toBeLessThan(5000);
    });
  });

  describe('Very large arrays - performance test', () => {
    it('should complete in reasonable time for 1000x1000 arrays', () => {
      const x = generateRandomValues(1000, 9999);
      const y = generateRandomValues(1000, 10000);
      const p = [0.5];

      const start = Date.now();
      const result = fastShift(x, y, p);
      const elapsed = Date.now() - start;

      expect(result.length).toBe(1);
      expect(elapsed).toBeLessThan(10000);
    });
  });

  describe('Many quantiles - performance test', () => {
    it('should handle many quantiles efficiently', () => {
      const x = generateRandomValues(200, 7777);
      const y = generateRandomValues(200, 7778);
      const p = Array.from({ length: 21 }, (_, i) => i / 20.0);

      const start = Date.now();
      const result = fastShift(x, y, p);
      const elapsed = Date.now() - start;

      expect(result.length).toBe(21);
      expect(elapsed).toBeLessThan(5000);
    });
  });

  describe('Null inputs - throws exception', () => {
    it('should throw error for null inputs', () => {
      const x = [1, 2];
      const y = [3, 4];
      const p = [0.5];

      expect(() => fastShift(null as unknown as number[], y, p)).toThrow();
      expect(() => fastShift(x, null as unknown as number[], p)).toThrow();
      expect(() => fastShift(x, y, null as unknown as number[])).toThrow();
    });
  });

  describe('Empty arrays - throws exception', () => {
    it('should throw error for empty arrays', () => {
      const empty: number[] = [];
      const valid = [1, 2];
      const p = [0.5];

      expect(() => fastShift(empty, valid, p)).toThrow('x and y must be non-empty');
      expect(() => fastShift(valid, empty, p)).toThrow('x and y must be non-empty');
    });
  });

  describe('Invalid probabilities - throws exception', () => {
    it('should throw error for invalid probabilities', () => {
      const x = [1, 2];
      const y = [3, 4];

      expect(() => fastShift(x, y, [-0.1])).toThrow('Probabilities must be within [0, 1]');
      expect(() => fastShift(x, y, [1.1])).toThrow('Probabilities must be within [0, 1]');
      expect(() => fastShift(x, y, [NaN])).toThrow('Probabilities must be within [0, 1]');
    });
  });

  describe('NaN in data - throws exception', () => {
    it('should throw error for NaN in data', () => {
      const xWithNaN = [1, NaN];
      const yWithNaN = [3, NaN];
      const valid = [1, 2];
      const p = [0.5];

      expect(() => fastShift(xWithNaN, valid, p)).toThrow('NaN values found in x');
      expect(() => fastShift(valid, yWithNaN, p)).toThrow('NaN values found in y');
    });
  });

  describe('Empty probabilities - returns empty', () => {
    it('should return empty array for empty probabilities', () => {
      const x = [1, 2];
      const y = [3, 4];
      const p: number[] = [];

      const result = fastShift(x, y, p);

      expect(result.length).toBe(0);
    });
  });

  describe('Shift invariance - X shift', () => {
    it('should be invariant to shifts in X', () => {
      const seed = 555;

      for (let trial = 0; trial < 5; trial++) {
        const x = generateRandomValues(15, seed + trial);
        const y = generateRandomValues(12, seed + trial + 100);
        const p = [0.25, 0.5, 0.75];
        const shift = trial * 10;

        const result1 = fastShift(x, y, p);
        const xShifted = x.map((v) => v + shift);
        const result2 = fastShift(xShifted, y, p);

        for (let i = 0; i < result1.length; i++) {
          expect(result2[i]).toBeCloseTo(result1[i] + shift, 9);
        }
      }
    });
  });

  describe('Shift invariance - Y shift', () => {
    it('should be invariant to shifts in Y', () => {
      const seed = 666;

      for (let trial = 0; trial < 5; trial++) {
        const x = generateRandomValues(15, seed + trial);
        const y = generateRandomValues(12, seed + trial + 100);
        const p = [0.25, 0.5, 0.75];
        const shift = trial * 10;

        const result1 = fastShift(x, y, p);
        const yShifted = y.map((v) => v + shift);
        const result2 = fastShift(x, yShifted, p);

        for (let i = 0; i < result1.length; i++) {
          expect(result2[i]).toBeCloseTo(result1[i] - shift, 9);
        }
      }
    });
  });

  describe('Scale invariance', () => {
    it('should scale appropriately with data', () => {
      const seed = 777;

      for (let trial = 0; trial < 5; trial++) {
        const x = generateRandomValues(15, seed + trial);
        const y = generateRandomValues(12, seed + trial + 100);
        const p = [0.5];
        const scale = 2.0;

        const result1 = fastShift(x, y, p);
        const xScaled = x.map((v) => v * scale);
        const yScaled = y.map((v) => v * scale);
        const result2 = fastShift(xScaled, yScaled, p);

        for (let i = 0; i < result1.length; i++) {
          expect(result2[i]).toBeCloseTo(result1[i] * scale, 6);
        }
      }
    });
  });
});
